src/main/kotlin/org/elm/ide/actions/ElmExternalFormatAction.kt

```diff
@@ -1,75 +1,69 @@
 
 
 class ElmExternalFormatAction : AnAction() {
 
+
     override fun update(e: AnActionEvent) {
         super.update(e)
         e.presentation.isEnabled = getContext(e) != null
     }
 
+    override fun actionPerformed(e: AnActionEvent) {
 
-    private fun getContext(e: AnActionEvent): Context? {
-        val project = e.project ?: return null
-        val toolchain = project.elmToolchain ?: return null
-        val file = e.getData(CommonDataKeys.PSI_FILE) ?: return null
-        if (!file.virtualFile.isInLocalFileSystem) return null
-        if (file !is ElmFile) return null
-        val elmVersion = when (val elmProject = file.elmProject) {
-            is ElmApplicationProject -> elmProject.elmVersion
-            is ElmPackageProject -> elmProject.elmVersion.low
-            else -> return null
+        val elmFormat = e.project?.elmToolchain?.elmFormat
+        if (elmFormat == null) {
+            e.project?.showBalloon("could not find elm-format", NotificationType.ERROR)
+            return
         }
-        return Context(project, toolchain, file, elmVersion)
-    }
 
+        val ctx = getContext(e)
 
-    override fun actionPerformed(e: AnActionEvent) {
-        val ctx = getContext(e) ?: return
-        val project = ctx.project
-        val virtualFile = ctx.elmFile.virtualFile
-        val elmFormat = ctx.toolchain.elmFormat
-        if (elmFormat == null) {
-            project.showBalloon("could not find elm-format", NotificationType.ERROR)
+        if (ctx == null) {
+            if (isUnitTestMode) throw RuntimeException("something went wrong initializing the context for elm-format")
             return
         }
 
-        FileDocumentManager.getInstance().saveAllDocuments()
         try {
-            ProgressManager.getInstance().runProcessWithProgressSynchronously<ProcessOutput, ExecutionException>({
-                elmFormat.reformatFile(project, ctx.elmVersion, virtualFile)
-            }, "Running elm-format on current file...", true, project)
-            // We want to refresh file synchronously only in unit test
-            // to get new text right after `reformatFile` call
-            VfsUtil.markDirtyAndRefresh(!isUnitTestMode, true, true, virtualFile)
-        } catch (e: ExecutionException) {
-            if (isUnitTestMode) throw e
-            val message = e.message ?: "something went wrong running elm-format"
-            project.showBalloon(message, NotificationType.ERROR)
+            elmFormat.formatDocumentAndSetText(ctx.project, ctx.document, ctx.elmVersion)
+        } catch (ex: ExecutionException) {
+            if (isUnitTestMode) throw ex
+            val message = ex.message ?: "something went wrong running elm-format"
+            e.project?.showBalloon(message, NotificationType.ERROR)
+
         }
     }
 
+    private fun getContext(e: AnActionEvent): Context? {
+        val project = e.project ?: return null
+        val file = e.getData(CommonDataKeys.PSI_FILE) ?: return null
```

Comment:
Now that we're using the VirtualFile, we don't actually need the PsiFile here at all. To get the virtual file: 
`val file = e.getData(CommonDataKeys.VIRTUAL_FILE) ?: return null`

Then to check that it's an Elm file, do `if (file.fileType != ElmFileType) return null`.